Add per-organization bucket name configuration

mtneug opened the pull request:
Add the ability to configure S3 buckets for each organization (asset manager and distribution). The old configuration is still compatible and will result in a default bucket configuration, i.e. if no org-specific buckets are configured, the default bucket will be used.

This likely depends on #6252 (I have not tested without this additional patch).

### How to test this patch

1. Configure two tenants (e.g. using https://github.com/opencast/community-integrations/tree/main/multi-tenant)
2. Configure different S3 buckets for each tenant
3. Upload videos to each tenant
4. Observe which S3 buckets are used

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://docs.opencast.org/develop/developer/#participate/development-process/#automatically-closing-issues-when-a-pr-is-merged) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)
* [ ] explain why it needs to be merged into the legacy branch, if it is targeting the legacy branch

mtneug committed:
Add per-organization bucket name configuration

Add the ability to configure S3 buckets for each organization
(asset manager and distribution). The old configuration is still
compatible and will result in a default bucket configuration,
i.e. if no org-specific buckets are configured, the default bucket
will be used.

This likely depends on #6252 (I have not tested without this
additional patch).

KatrinIhler commented:
ICLA check still fails, wtf?

KatrinIhler commented:
FYI: Branch cut for OC 17 is on November 6!

gregorydlogan commented on modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/s3/AwsS3AssetStore.java:
```diff
@@ -180,9 +186,21 @@ public void activate(final ComponentContext cc) throws IllegalStateException, Co
+            orgBucketNameMap.put(orgId, bucketName);
+          });
 
       // AWS region
```
So what if I don't define *any* buckets by accident?  Can we check at this point to make sure at least one bucket definition exists?  The previous behaviour of the `createAWSBucket` method (poorly) enforced this by throwing a NPE when `s3.listObjects(null)` was called :)

gregorydlogan commented on modules/distribution-service-aws-s3/src/main/java/org/opencastproject/distribution/aws/s3/AwsS3DistributionServiceImpl.java:
```diff
@@ -257,9 +262,21 @@ public void activate(ComponentContext cc) {
+            orgBucketNameMap.put(orgId, bucketName);
+          });
 
       // AWS region
```
Similar complaint to above, lets verify that at least one is set.

gregorydlogan commented on modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/s3/AwsS3AssetStore.java:
```diff
@@ -544,12 +572,24 @@ private void restoreGlacierObject(String objectName, Integer objectRestorePeriod
     }
   }
 
+  private String getBucketName(String orgId) {
```
This needs release notes.  Users with existing S3 configs *must* ensure that the default bucket is still available (either as the default, or as an org bucket) otherwise the data won't be accessible since this method won't return the right bucket name

mtneug replied:
Users with unchanged config use `org.opencastproject.assetmanager.aws.s3.bucket`. This is now stored in the map as the default bucket. I don't think users need to change anything if the current behavior is good enough for them.

gregorydlogan commented on modules/distribution-service-aws-s3/src/main/java/org/opencastproject/distribution/aws/s3/AwsS3DistributionServiceImpl.java:
```diff
@@ -994,6 +1021,18 @@ protected void setStorageTmp(String path) {
     }
   }
 
+  private String getBucketName(String orgId) {
```
Same comment about release notes.

mtneug replied:
Same implementation with the `org.opencastproject.distribution.aws.s3.bucket` config key.

mtneug committed:
Add check to ensure buckets are configured

mtneug commented:
@gregorydlogan I rebased and added the checks for bucket configuration. As for the release notes, I can add a line, but this change should not be breaking with existing configuration.

gregorydlogan commented:
If we don't need to alter configs to keep things working then IMO this is just a new feature and doens't need the notes.
